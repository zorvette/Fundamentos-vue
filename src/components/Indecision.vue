<template>
  <!-- <img src="https://stickerly.pstatic.net/sticker_pack/4LRevtriqE5N13OuYyntUg/0EMLYD/3/62304888-19f6-4ecc-a740-89144a388e48.png" alt="bg"> -->
  <img v-if="img" :src="img" alt="bg">

  <div class="bg-dark"></div>

  <div class="indecision-container">

    <input 
    type="text" 
    placeholder="Hazme una pregunta" 
    v-model="question"
    >

    <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>

    <div v-if="isValidQuestion">
      <h2>{{question}}</h2>
      <h1>{{ answer }}</h1>
    </div>

  </div>

</template>


<script>
export default {
  data() {
    return{
      question: null,
      answer: null,
      img: null,
      isValidQuestion: false
    }
  },
  methods: {
    async getAnswer() {

      try {
        this.answer = 'Pensando...'
        const {answer, image} = await fetch ('https://yesno.wtf/api').then( r => r.json() )
        
        // console.log(answer)
        this.answer = answer === 'yes' ? 'Si!' : 'No!'
        this.img = image

      } catch (error) {
        console.log( 'Indecision Componet', error )
        this.answer = 'No se pudo cargar del API'
        this.img = null

      }
    }
  },
  watch: {
    question( value, oldValue ) {

      this.isValidQuestion = false

      console.log({ value })

      if( !value.includes('?') ) return
      console.log({ value })

      this.isValidQuestion = true

      // TODO: Realizar peticion http
      this.getAnswer()
    }
  }
  }
</script>

<style>

  img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 0px;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }

</style>