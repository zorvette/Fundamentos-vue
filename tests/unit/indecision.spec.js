import { shallowMount } from '@vue/test-utils'
import Indecision from '@/components/Indecision'

describe('Indecision Component', () => {

  let wrapper
  let clgSpy
  let getAnswerSpy

  global.fetch = jest.fn( () => Promise.resolve( {
    json: () => Promise.resolve({ 
      answer: 'yes',
      forced: false,
      image: 'https://yesno.wtf/assets/yes/3-422e51268d64d78241720a7de52fe121.gif'
     })
  }))

  beforeEach (() => {
    wrapper = shallowMount( Indecision )

    clgSpy = jest.spyOn( console, 'log' )

    getAnswerSpy = jest.spyOn ( wrapper.vm, 'getAnswer')
    
    jest.clearAllMocks()
  })


  test ('debe hacer match con el snapshot', () => {
    expect( wrapper.html() ).toMatchSnapshot()
  })

  test('escribir en el input no debe de disparar nada (console.log)', async() => {

    const input = wrapper.find('input')
    await input.setValue('Hola Mundo')

    expect( clgSpy ).toHaveBeenCalledTimes(1)
    expect( getAnswerSpy ).not.toHaveBeenCalled()

    // console.log(wrapper.vm)
  })

  test('escribir el simbolo "?" debe de disparar el getAnswer', async () => {

    const input = wrapper.find('input')
    await input.setValue('Hola Mundo?')

    expect( clgSpy ).toHaveBeenCalledTimes(2)
    expect( getAnswerSpy ).toHaveBeenCalled()

  })

  test('pruebas en getAnswer', async()=> {

    await wrapper.vm.getAnswer()

    const img = wrapper.find('img')
    
    expect( img.exists() ).toBeTruthy()
    expect( wrapper.vm.img ).toBe('https://yesno.wtf/assets/yes/3-422e51268d64d78241720a7de52fe121.gif')
    expect( wrapper.vm.answer ).toBe('Si!')

  })

  test('prueba en getAnswer - Fallo en el API', async() => {

    fetch.mockImplementationOnce( () => Promise.reject(' API is down ') )

    await wrapper.vm.getAnswer()

    const img = wrapper.find('img')

    expect( wrapper.vm.img ).toBeFalsy()
    expect( wrapper.vm.answer ).toBe('No se pudo cargar del API')

  })
})