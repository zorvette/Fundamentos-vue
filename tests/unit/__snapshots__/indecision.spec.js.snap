// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component debe hacer match con el snapshot 1`] = `
<!-- <img src="https://stickerly.pstatic.net/sticker_pack/4LRevtriqE5N13OuYyntUg/0EMLYD/3/62304888-19f6-4ecc-a740-89144a388e48.png" alt="bg"> -->
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
  <!--v-if-->
</div>
`;
